contrib/mpi-proxy-split/split_process.h

```diff
@@ -76,12 +79,36 @@ static inline void SET_LOWER_HALF_FS_CONTEXT() {
     syscall(SYS_arch_prctl, ARCH_GET_FS, &uh_fsaddr);
     uh_fsaddr = uh_fsaddr - BUF_SIZE;
   }
-  memcpy(fsaddr_buf, uh_fsaddr, BUF_SIZE + 3*sizeof(void *)); // & 3 ptrs of TCB
-  memcpy(uh_fsaddr, lh_fsaddr, BUF_SIZE + 3*sizeof(void *));
+  memcpy(fsaddr_buf, uh_fsaddr, BUF_SIZE + TCB_HEADER_SIZE);
+  memcpy(uh_fsaddr, lh_fsaddr, BUF_SIZE + TCB_HEADER_SIZE);
+
+  // on x86_64:
+  // the tcb starts with 3 ptrs: self, dtv, header
+  // self is used to get TLS variables
+  // dtv is used to load tls for dynamically loaded libraries
+  // header is used to access the tcb
+  // self, header, and fs reg all point to the start of the tcb
+
+  // we change the self pointed to point to the tcb at the new location
+  // we leave the header pointer intact so that the tcb is modified in-place
+  // at the old location
+
+  // changing the location breaks things that keep pointers into the struct,
+  // like linked lists. TCB has a lot of internal self-pointers and lists, so
+  // it is advantageous to keep it in place. TLS usually has less (application
+  // dependent), and must be moved because it can be accessed relative to fs
+  // with no indirection. So we change self but not header.
+
+  // change self pointer to new application half TLS location
+  ((void **)(uh_fsaddr + BUF_SIZE))[0] = (void *) (uh_fsaddr + BUF_SIZE);
```

Comment:
Don't we need to update the other two pointers -- dtv and header?

Reply:
I think we don't have to
For dtv, because mana statically linked libs in the lower half, so this won't change, if mana will use dynamical linking in the future, link it with `-ftls-model=initial-exec` can also avoid the problem.
For `self`, we can let it access the TCB in the old place, we may don't need to copy back TCB when restore, let me make the change and give it another try.